<a href="https://colab.research.google.com/github/hamant-jagwan/Hotel-Booking--EDA/blob/main/hotel_booking_EDA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# FIRST VIEW OF DATASET

importing libraires


In [48]:
# importing the libraries
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
# configuration
pd.set_option('display.max_columns', None) #sets the maximum number of columns to be displayed when printing a Pandas DataFrame

In [49]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [50]:
file_path = '/content/drive/MyDrive/data/hotel_booking.csv'
hb_df = pd.read_csv(file_path)

## DATA FIRST VIEW

In [51]:
hb_df.head(5) # view of hotel booking data

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  \
0                        0                     0       2       0.0       0   
1                        0                     0       2       0.0       0   
2                        0                     1       1       0.0       0   
3                        0                     1       1       0.0       0   
4                        0                     2       2       0.0       0   

  meal country market_segment distribution_channel  is_repeated_guest  \
0   BB     PRT         Direct               Direct                  0   
1   BB     PRT         Direct               Direct                  0   
2   BB     GBR         Direct               Direct                  0   
3   BB     GBR      Corporate            Corporate                  0   
4   BB     GBR      Online TA                TA/TO                  0   

   previous_cancellations  previous_bookings_not_canceled reserved_room_type  \
0                       0                               0                  C   
1                       0                               0                  C   
2                       0                               0                  A   
3                       0                               0                  A   
4                       0                               0                  A   

  assigned_room_type  booking_changes deposit_type  agent  company  \
0                  C                3   No Deposit    NaN      NaN   
1                  C                4   No Deposit    NaN      NaN   
2                  C                0   No Deposit    NaN      NaN   
3                  A                0   No Deposit  304.0      NaN   
4                  A                0   No Deposit  240.0      NaN   

   days_in_waiting_list customer_type   adr  required_car_parking_spaces  \
0                     0     Transient   0.0                            0   
1                     0     Transient   0.0                            0   
2                     0     Transient  75.0                            0   
3                     0     Transient  75.0                            0   
4                     0     Transient  98.0                            0   

   total_of_special_requests reservation_status reservation_status_date  \
0                          0          Check-Out              2015-07-01   
1                          0          Check-Out              2015-07-01   
2                          0          Check-Out              2015-07-02   
3                          0          Check-Out              2015-07-02   
4                          1          Check-Out              2015-07-03   

             name                        email  phone-number       credit_card  
0   Ernest Barnes  Ernest.Barnes31@outlook.com  669-792-1661  ************4322  
1    Andrea Baker       Andrea_Baker94@aol.com  858-637-6955  ************9157  
2  Rebecca Parker   Rebecca_Parker@comcast.net  652-885-2745  ************3734  
3    Laura Murray            Laura_M@gmail.com  364-656-8427  ************5677  
4     Linda Hines           LHines@verizon.com  713

In [52]:
hb_df.tail()

hotel  is_canceled  lead_time  arrival_date_year  \
119385  City Hotel            0         23               2017   
119386  City Hotel            0        102               2017   
119387  City Hotel            0         34               2017   
119388  City Hotel            0        109               2017   
119389  City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  children  babies meal country  \
119385                     5       2       0.0       0   BB     BEL   
119386                     5       3       0.0       0   BB     FRA   
119387                     5       2       0.0       0   BB     DEU   
119388                     5       2       0.0       0   BB     GBR   
119389                     7       2       0.0       0   HB     DEU   

       market_segment distribution_channel  is_repeated_guest  \
119385  Offline TA/TO                TA/TO                  0   
119386      Online TA                TA/TO                  0   
119387      Online TA                TA/TO                  0   
119388      Online TA                TA/TO                  0   
119389      Online TA                TA/TO                  0   

        previous_cancellations  previous_bookings_not_canceled  \
119385                       0                               0   
119386                       0                               0   
119387                       0                               0   
119388                       0                               0   
119389                       0                               0   

       reserved_room_type assigned_room_type  booking_changes deposit_type  \
119385                  A                  A                0   No Deposit   
119386                  E                  E                0   No Deposit   
119387                  D                  D                0   No Deposit   
119388                  A                  A                0   No Deposit   
119389                  A                  A                0   No Deposit   

        agent  company  days_in_waiting_list customer_type     adr  \
119385  394.0      NaN                     0     Transient   96.14   
119386    9.0      NaN                     0     Transient  225.43   
119387    9.0      NaN                     0     Transient  157.71   
119388   89.0      NaN                     0     Transient  104.40   
119389    9.0      NaN                     0     Transient  151.20   

        required_car_parking_spaces  total_of_special_requests  \
119385                            0                          0   
119386                            0                          2   
119387                            0                          4   
119388                            0                          0   
119389                            0                          2   

       reservation_status reservation_status_date                name  \
119385          Check-Out              2017-09-06     Claudia Johnson   
119386          Check-Out              2017-09-07      Wesley Aguilar   
119387          Check-Out              2017-09-07        Mary Morales   
119388          Check-Out              2017-09-07  Caroline Conley MD   
119389          Check-Out              2017-09-07  

In [53]:
# removing last 4 columns because it is not neccesary for analysis  and contain personal information
hb_df = hb_df.iloc[:,:-4]

In [54]:
hb_df.shape

(119390, 32)

In [55]:
hb_df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [56]:
hb_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

In [57]:
# This user define function will give the type,count of null and non null values as well as null ratio
def hotelbookinginfo():
  temp=pd.DataFrame(index=hb_df.columns)
  temp["datatype"]=hb_df.dtypes
  temp["not null values"]=hb_df.count()
  temp["null value"]=hb_df.isnull().sum()
  temp["% of the null value"]=hb_df.isnull().mean()*100
  temp["unique count"]=hb_df.nunique() # Return Series with number of distinct elements. Can ignore NaN values.
  return temp
hotelbookinginfo()


datatype  not null values  null value  \
hotel                            object           119390           0   
is_canceled                       int64           119390           0   
lead_time                         int64           119390           0   
arrival_date_year                 int64           119390           0   
arrival_date_month               object           119390           0   
arrival_date_week_number          int64           119390           0   
arrival_date_day_of_month         int64           119390           0   
stays_in_weekend_nights           int64           119390           0   
stays_in_week_nights              int64           119390           0   
adults                            int64           119390           0   
children                        float64           119386           4   
babies                            int64           119390           0   
meal                             object           119390           0   
country                          object           118902         488   
market_segment                   object           119390           0   
distribution_channel             object           119390           0   
is_repeated_guest                 int64           119390           0   
previous_cancellations            int64           119390           0   
previous_bookings_not_canceled    int64           119390           0   
reserved_room_type               object           119390           0   
assigned_room_type               object           119390           0   
booking_changes                   int64           119390           0   
deposit_type                     object           119390           0   
agent                           float64           103050       16340   
company                         float64             6797      112593   
days_in_waiting_list              int64           119390           0   
customer_type                    object           119390           0   
adr                             float64           119390           0   
required_car_parking_spaces       int64           119390           0   
total_of_special_requests         int64           119390           0   
reservation_status               object           119390           0   
reservation_status_date          object           119390           0   

                                % of the null value  unique count  
hotel                                      0.000000             2  
is_canceled                                0.000000             2  
lead_time                                  0.000000           479  
arrival_date_year                          0.000000             3  
arrival_date_month                         0.000000            12  
arrival_date_week_number                   0.000000            53  
arrival_date_day_of_month                  0.000000            31  
stays_in_weekend_nights                    0.000000            17  
stays_in_week_nights                       0.000000            35  
adults                                     0.000000            14  
children                                   0.003350             5  
babies                                     0.000000             5  
meal                                       0.000000             5  
country                                    0.408744           177  
market_segment                             0.000000             8  
distribution_channel                       0.000000             5  
is_repeated_guest                          0.000000             2  
previous_cancellations                     0.000000            15  
previous_bookings_not_canceled             0.000000            73  
reserved_room_type                         0.000000            10  
assigned_room_type                         0.000000            12  
booking_changes                            0.000000            21  
deposit_type                               0.000000             3  
agent                 

The columns and the data it represents are listed below:

**1. hotel :** Name of the hotel (Resort Hotel or City Hotel)

**2. is_canceled :** If the booking was canceled (1) or not (0)

**3. lead_time :** Number of days before the actual arrival of the guests

**4. arrival_date_year :** Year of arrival date

**5. arrival_date_month :** Month of month arrival date

**6. arrival_date_week_number :** Week number of year for arrival date

**7. arrival_date_day_of_month :** Day of arrival date

**8. stays_in_weekend_nights :** Number of weekend nights (Saturday or Sunday) spent at the hotel by the guests.

**9. stays_in_week_nights :** Number of weeknights (Monday to Friday) spent at the hotel by the guests.

**10. adults :** Number of adults among guests

**11. children :** Number of children among guests

**12. babies :** Number of babies among guests

**13. meal :** Type of meal booked

**14. country :** Country of guests

**15. market_segment :** Designation of market segment

**16. distribution_channel :** Name of booking distribution channel

**17. is_repeated_guest :** If the booking was from a repeated guest (1) or not (0)

**18. previous_cancellations :** Number of previous bookings that were cancelled by the customer prior to the current booking

**19. previous_bookings_not_canceled :** Number of previous bookings not cancelled by the customer prior to the current booking

**20. reserved_room_type :** Code of room type reserved

**21. assigned_room_type :** Code of room type assigned

**22. booking_changes :** Number of changes/amendments made to the booking

**23. deposit_type :** Type of the deposit made by the guest

**24. agent :** ID of travel agent who made the booking

**25. company :** ID of the company that made the booking

**26. days_in_waiting_list :** Number of days the booking was in the waiting list

**27. customer_type :** Type of customer, assuming one of four categories

**28. adr :** Average Daily Rate, as defined by dividing the sum of all lodging transactions by the total number of staying nights

**29. required_car_parking_spaces :** Number of car parking spaces required by the customer

**30. total_of_special_requests :** Number of special requests made by the customer

**31. reservation_status :** Reservation status (Canceled, Check-Out or No-Show)

**32. reservation_status_date :** Date at which the last reservation status was updated

In [58]:
# converting the `reservation_status_date` into datetime datatype

hb_df['reservation_status_date'] = pd.to_datetime(hb_df['reservation_status_date'])

In [59]:
hb_df.describe(include = 'object').T

count unique         top    freq
hotel                 119390      2  City Hotel   79330
arrival_date_month    119390     12      August   13877
meal                  119390      5          BB   92310
country               118902    177         PRT   48590
market_segment        119390      8   Online TA   56477
distribution_channel  119390      5       TA/TO   97870
reserved_room_type    119390     10           A   85994
assigned_room_type    119390     12           A   74053
deposit_type          119390      3  No Deposit  104641
customer_type         119390      4   Transient   89613
reservation_status    119390      3   Check-Out   75166

In [60]:
for col in hb_df.describe(include = 'object').columns:
  print(col)

  print(hb_df[col].unique())
  print("-" *80)

hotel
['Resort Hotel' 'City Hotel']
--------------------------------------------------------------------------------
arrival_date_month
['July' 'August' 'September' 'October' 'November' 'December' 'January'
 'February' 'March' 'April' 'May' 'June']
--------------------------------------------------------------------------------
meal
['BB' 'FB' 'HB' 'SC' 'Undefined']
--------------------------------------------------------------------------------
country
['PRT' 'GBR' 'USA' 'ESP' 'IRL' 'FRA' nan 'ROU' 'NOR' 'OMN' 'ARG' 'POL'
 'DEU' 'BEL' 'CHE' 'CN' 'GRC' 'ITA' 'NLD' 'DNK' 'RUS' 'SWE' 'AUS' 'EST'
 'CZE' 'BRA' 'FIN' 'MOZ' 'BWA' 'LUX' 'SVN' 'ALB' 'IND' 'CHN' 'MEX' 'MAR'
 'UKR' 'SMR' 'LVA' 'PRI' 'SRB' 'CHL' 'AUT' 'BLR' 'LTU' 'TUR' 'ZAF' 'AGO'
 'ISR' 'CYM' 'ZMB' 'CPV' 'ZWE' 'DZA' 'KOR' 'CRI' 'HUN' 'ARE' 'TUN' 'JAM'
 'HRV' 'HKG' 'IRN' 'GEO' 'AND' 'GIB' 'URY' 'JEY' 'CAF' 'CYP' 'COL' 'GGY'
 'KWT' 'NGA' 'MDV' 'VEN' 'SVK' 'FJI' 'KAZ' 'PAK' 'IDN' 'LBN' 'PHL' 'SEN'
 'SYC' 'AZE' 'BHR' 'NZL' 'THA' 'DO

In [61]:
hb_df.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

In [62]:
hb_df.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In the above data we have four columns that have null values and these columns are the following

1. `children`  = 4
2. `country`   = 488
3. `agent`     = 16340
4. `company`   = 112593

* There are some missing values in the `children` and `country` fields. We can safely remove these values from the dataset.

* We can remove the `agent` and `company` columns due to excessive null values, as they are not necessary for our analysis.

In [63]:
hb_df.drop(['company', 'agent'], axis = 1 , inplace = True) #In order to remove a column, it is imperative to specify axis=1.
hb_df.dropna(inplace = True)

In [65]:
# checking the value
hb_df.isnull().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          0
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   